File: scripts/generate-hero-video.ts
/**
 * scripts/generate-hero-video.ts · Wave 3l · Veo 3.1 hero video generator.
 *
 * Calls Google Veo 3.1 (via the Gemini API) to generate an atmospheric
 * nature loop for the home hero. Polls for completion (Veo is async;
 * generations take ~30-90s), downloads the returned MP4, and drops it
 * at `public/hero-video.mp4`.
 *
 * Usage:
 *   # 1. Get a Gemini API key from https://ai.google.dev/gemini-api
 *   # 2. Set it in your environment:
 *   export GEMINI_API_KEY="AIza…"
 *   # 3. Run:
 *   npx tsx scripts/generate-hero-video.ts
 *
 * Optional args:
 *   --prompt "..."     override the default prompt
 *   --count N          how many variants (default 1)
 *   --out path.mp4     output path (default public/hero-video.mp4)
 *
 * After a video lands:
 *   • Extract a poster still (first frame) via ffmpeg:
 *       ffmpeg -i public/hero-video.mp4 -vf "select=eq(n\,0)" -q:v 2 public/hero-poster.jpg
 *   • The HomeHero component picks up both files automatically on next build.
 *
 * Cost: ~$3-5 per 8s clip at 1080p (Veo 3.1 preview pricing).
 * Model: `veo-3.1-generate-preview` (or bump when the GA release lands).
 */

import { GoogleGenAI } from "@google/genai";
import fs from "node:fs";
import path from "node:path";

const DEFAULT_PROMPT = `Ambient landscape · drifting fog rolls slowly over a single mountain peak at first light · warm-cool color balance · muted ochre and slate palette · subtle 35mm film grain · silent · no cuts · one continuous shot · 8 seconds · horizontal composition · natural-history-monograph aesthetic · quiet institutional beauty · editorial cinematography · no text · no people · no logos.`;

const DEFAULT_MODEL = "veo-3.1-generate-preview";
const DEFAULT_OUT = "public/hero-video.mp4";

type Args = { prompt: string; count: number; out: string; model: string };

function parseArgs(argv: string[]): Args {
  const args: Args = { prompt: DEFAULT_PROMPT, count: 1, out: DEFAULT_OUT, model: DEFAULT_MODEL };
  for (let i = 0; i < argv.length; i++) {
    const flag = argv[i];
    const next = argv[i + 1];
    if (flag === "--prompt" && next) { args.prompt = next; i++; }
    else if (flag === "--count" && next) { args.count = Number(next); i++; }
    else if (flag === "--out" && next) { args.out = next; i++; }
    else if (flag === "--model" && next) { args.model = next; i++; }
  }
  return args;
}

async function main() {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) {
    console.error("[generate-hero-video] Missing GEMINI_API_KEY environment variable.");
    console.error("  Get a key at https://ai.google.dev/gemini-api then:");
    console.error("    export GEMINI_API_KEY=\"AIza…\"");
    process.exit(1);
  }

  const args = parseArgs(process.argv.slice(2));
  const client = new GoogleGenAI({ apiKey });

  console.log(`[generate-hero-video] prompt:`);
  console.log(`  ${args.prompt}`);
  console.log(`[generate-hero-video] model: ${args.model}`);
  console.log(`[generate-hero-video] variants: ${args.count}`);

  for (let variant = 1; variant <= args.count; variant++) {
    const outPath = args.count > 1
      ? args.out.replace(/(\.mp4)$/i, `-v${variant}$1`)
      : args.out;

    console.log(`\n[variant ${variant}/${args.count}] starting generation → ${outPath}`);

    // Kick off generation (returns an operation immediately)
    let operation = await client.models.generateVideos({
      model: args.model,
      prompt: args.prompt,
      config: {
        aspectRatio: "16:9",
        durationSeconds: 8,
        numberOfVideos: 1,
      },
    });

    // Poll until the operation completes (~30-90s typical)
    const startedAt = Date.now();
    while (!operation.done) {
      await new Promise((r) => setTimeout(r, 5000));
      operation = await client.operations.getVideosOperation({ operation });
      const elapsed = Math.round((Date.now() - startedAt) / 1000);
      console.log(`[variant ${variant}] polling · ${elapsed}s elapsed · done=${operation.done}`);
      if (elapsed > 300) {
        console.error(`[variant ${variant}] TIMEOUT after 300s`);
        break;
      }
    }

    const videos = operation.response?.generatedVideos ?? [];
    if (!videos.length) {
      console.error(`[variant ${variant}] no video returned in operation response`);
      console.error(JSON.stringify(operation, null, 2).slice(0, 1200));
      continue;
    }

    // Download the first (only) video
    const video = videos[0]!;
    console.log(`[variant ${variant}] downloading video…`);
    await client.files.download({
      file: video.video!,
      downloadPath: outPath,
    });

    const stat = fs.statSync(outPath);
    console.log(`[variant ${variant}] ✓ saved ${outPath} (${(stat.size / 1024 / 1024).toFixed(2)} MB)`);
  }

  console.log(`\n[generate-hero-video] done.`);
  console.log(`\nNext steps:`);
  console.log(`  1. Extract a poster still:`);
  console.log(`     ffmpeg -i ${args.out} -vf "select=eq(n\\,0)" -q:v 2 public/hero-poster.jpg`);
  console.log(`  2. Verify locally: npm run dev, open http://localhost:3000`);
  console.log(`  3. Commit both files (git add public/hero-video.mp4 public/hero-poster.jpg)`);
}

main().catch((err) => {
  console.error("[generate-hero-video] failed:", err?.message ?? err);
  if (err?.stack) console.error(err.stack);
  process.exit(1);
});

File: scripts/generate-hero-video.py
#!/usr/bin/env -S uv run --with google-genai --python 3.12
"""
scripts/generate-hero-video.py · Wave 3l · Veo 3.1 hero video generator.

Uses Google's current unified Interactions API (`client.interactions.create`)
via the google-genai Python SDK. Founder pasted the exact API shape they
want us to use — this script mirrors it.

Usage:
    # 1. Get a Gemini API key: https://ai.google.dev/gemini-api
    # 2. Export it:
    export GEMINI_API_KEY="AIza…"
    # 3. Run (uv handles the SDK install per-run — no `uv add` needed):
    uv run scripts/generate-hero-video.py
    # or with args:
    uv run scripts/generate-hero-video.py --count 3

Optional args:
  --prompt "..."     override the default prompt
  --count N          how many variants (default 1)
  --out path.mp4     output path (default public/hero-video.mp4)
  --model NAME       Veo model to use (default veo-3.1-generate-preview)

After a video lands:
  ffmpeg -i public/hero-video.mp4 -vf "select=eq(n\\,0)" -q:v 2 public/hero-poster.jpg

Cost: ~$3-5 per 8-second 1080p clip at Veo 3.1 preview pricing.
"""
from __future__ import annotations

import argparse
import os
import sys
import time
import urllib.request
from pathlib import Path

from google import genai
from google.genai import types

DEFAULT_PROMPT = (
    "Ambient landscape · drifting fog rolls slowly over a single mountain peak "
    "at first light · warm-cool color balance · muted ochre and slate palette · "
    "subtle 35mm film grain · silent · no cuts · one continuous shot · 8 seconds · "
    "horizontal composition · natural-history-monograph aesthetic · "
    "quiet institutional beauty · editorial cinematography · "
    "no text · no people · no logos."
)

DEFAULT_MODEL = "veo-3.1-generate-preview"
DEFAULT_OUT = "public/hero-video.mp4"


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Generate the nebbos.ai hero video via Veo 3.1.")
    p.add_argument("--prompt", default=DEFAULT_PROMPT, help="Veo prompt.")
    p.add_argument("--count", type=int, default=1, help="How many variants to generate.")
    p.add_argument("--out", default=DEFAULT_OUT, help="Output MP4 path.")
    p.add_argument("--model", default=DEFAULT_MODEL, help="Veo model id.")
    return p.parse_args()


def _try_interactions_create(client: genai.Client, model: str, prompt: str):
    """
    Prefer the unified Interactions API (founder-pasted pattern):
        client.interactions.create(model=..., input=..., ...)

    Fall back to the older models.generate_videos if this SDK build doesn't
    yet route video generation through Interactions. Keeps the script
    working across genai SDK versions.
    """
    try:
        return client.interactions.create(
            model=model,
            input=prompt,
            config=types.GenerateVideosConfig(
                aspect_ratio="16:9",
                duration_seconds=8,
                number_of_videos=1,
            ),
        ), "interactions"
    except AttributeError:
        # Older SDK — fall back to models.generate_videos
        return client.models.generate_videos(
            model=model,
            prompt=prompt,
            config=types.GenerateVideosConfig(
                aspect_ratio="16:9",
                duration_seconds=8,
                number_of_videos=1,
            ),
        ), "models"


def _download_video(client: genai.Client, video_obj, out_path: str) -> None:
    """Download the generated video to disk. Supports both response shapes."""
    Path(out_path).parent.mkdir(parents=True, exist_ok=True)
    # Newer SDK: has a .uri that requires the API key for download
    uri = getattr(video_obj, "uri", None)
    if uri:
        req = urllib.request.Request(uri, headers={"x-goog-api-key": os.environ["GEMINI_API_KEY"]})
        with urllib.request.urlopen(req) as r, open(out_path, "wb") as f:
            f.write(r.read())
        return
    # Older SDK: use client.files.download
    client.files.download(file=video_obj, download_path=out_path)


def main() -> int:
    api_key = os.getenv("GEMINI_API_KEY")
    if not api_key:
        print("[generate-hero-video] Missing GEMINI_API_KEY environment variable.", file=sys.stderr)
        print("  Get a key at https://ai.google.dev/gemini-api then:", file=sys.stderr)
        print('    export GEMINI_API_KEY="AIza…"', file=sys.stderr)
        return 1

    args = parse_args()
    client = genai.Client()  # picks up GEMINI_API_KEY from env

    print("[generate-hero-video] prompt:")
    print(f"  {args.prompt}")
    print(f"[generate-hero-video] model:  {args.model}")
    print(f"[generate-hero-video] count:  {args.count}")

    for variant in range(1, args.count + 1):
        out_path = (
            args.out
            if args.count == 1
            else str(Path(args.out).with_stem(Path(args.out).stem + f"-v{variant}"))
        )
        print(f"\n[variant {variant}/{args.count}] starting generation → {out_path}")

        operation, api = _try_interactions_create(client, args.model, args.prompt)
        print(f"[variant {variant}] using API surface: {api}")

        # Poll until the operation completes.
        started = time.time()
        while not operation.done:
            time.sleep(5)
            operation = client.operations.get(operation)
            elapsed = int(time.time() - started)
            print(f"[variant {variant}] polling · {elapsed}s elapsed · done={operation.done}")
            if elapsed > 300:
                print(f"[variant {variant}] TIMEOUT after 300s", file=sys.stderr)
                break

        response = getattr(operation, "response", None)
        videos = (
            getattr(response, "generated_videos", None)
            or getattr(response, "videos", None)
            or []
        )
        if not videos:
            print(f"[variant {variant}] no video returned in operation response", file=sys.stderr)
            print(repr(operation)[:1200], file=sys.stderr)
            continue

        video = videos[0]
        video_obj = getattr(video, "video", None) or video  # both shapes
        print(f"[variant {variant}] downloading video…")
        _download_video(client, video_obj, out_path)

        size_mb = Path(out_path).stat().st_size / 1024 / 1024
        print(f"[variant {variant}] ✓ saved {out_path} ({size_mb:.2f} MB)")

    print("\n[generate-hero-video] done.\n")
    print("Next steps:")
    if args.count > 1:
        print("  1. Pick a variant and rename to public/hero-video.mp4")
    print("  2. Extract a poster still:")
    print(f'     ffmpeg -i {args.out} -vf "select=eq(n\\,0)" -q:v 2 public/hero-poster.jpg')
    print("  3. Verify locally: npm run dev, open http://localhost:3000")
    print("  4. Commit: git add public/hero-video.mp4 public/hero-poster.jpg && git push")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
